src/keys.c

```diff
@@ -2934,12 +2934,23 @@ int DeriveKeys(WOLFSSL* ssl)
         if (keyData)   XFREE(keyData,   NULL, DYNAMIC_TYPE_TMP_BUFFER);
         if (md5)       XFREE(md5,       NULL, DYNAMIC_TYPE_TMP_BUFFER);
         if (sha)       XFREE(sha,       NULL, DYNAMIC_TYPE_TMP_BUFFER);
-        
+
         return MEMORY_E;
     }
 #endif
 
-    wc_InitMd5(md5);
+    ret = wc_InitMd5(md5);
+    if (ret != 0) {
+    #ifdef WOLFSSL_SMALL_STACK
+        XFREE(shaOutput, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```

Comment:
Too much code is added here.  Notice that other failures drop down to bottom to free in one place.